### ISSUE TITLE
overlapping issue about backward of LayerNormLinear

### ISSUE THREAD

**Issue by cos120** [NVIDIA/TransformerEngine#1353]
Hi, folks.

I am using 4 A100-SXM4 for with pytorch2.4.0 and mcore0.9.0 with transformer engine(0.11.0+fc03478) with tp2/pp2 and sequence parallel.

I found that if i set `TORCH_NCCL_ENABLE_TIMING=1` for timing all nccl operations, the ag/rs of sp in LayerNormLinear and LayerNormMLP will not overlap with dgrad/wgrad.

timeline without `TORCH_NCCL_ENABLE_TIMING=1`

there are 4 `cudaEventRecord`, but those events should create without timing flag.

![Image](https://github.com/user-attachments/assets/b4d42fe7-0d51-4458-a6c2-135767ab04ec)

![Image](https://github.com/user-attachments/assets/75f225a3-aa13-40be-8179-f381f37a84c6)

timeline with `TORCH_NCCL_ENABLE_TIMING=1`

the there are 5 `cudaEventRecord`, torch will let two of them create with timing flag.

![Image](https://github.com/user-attachments/assets/6031de87-a831-4a93-973f-68baf4ac2a63)

![Image](https://github.com/user-attachments/assets/46b7e5d2-b4dc-4c18-8f68-6ed982d707d7)

Why event record with timing will break the overlapping??  nccl operators use 24 sm, the matmul has enough space to launch.

![Image](https://github.com/user-attachments/assets/ae3ac701-01a6-4f92-a4bf-6a886855cb3e)

Here are timeline files, `torch_record.json` with `TORCH_NCCL_ENABLE_TIMING=1` and `no_record.json` does not set `TORCH_NCCL_ENABLE_TIMING`
[timeline.tar.gz](https://github.com/user-attachments/files/18001543/timeline.tar.gz)

**Comment by timmoon10** [comment]
It's strange that `TORCH_NCCL_ENABLE_TIMING=1` has this effect. I haven't been able to fully dig into what it does in PyTorch, but as far as I can tell the only extra thing is that it records a CUDA event before each NCCL collective: https://github.com/pytorch/pytorch/blob/e499b46465bc6e5f1a95f158e44bbf0f8356a220/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L2997
However, it's possible the profiler could have more complicated interactions.

Can you provide a timeline of the `LayerNormLinear` backward with `TORCH_NCCL_ENABLE_TIMING=1`? The provided timeline shows `Linear`, which does not overlap its tensor-parallel communication.

**Comment by cos120** [comment]
> It's strange that `TORCH_NCCL_ENABLE_TIMING=1` has this effect. I haven't been able to fully dig into what it does in PyTorch, but as far as I can tell the only extra thing is that it records a CUDA event before each NCCL collective: https://github.com/pytorch/pytorch/blob/e499b46465bc6e5f1a95f158e44bbf0f8356a220/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L2997
> However, it's possible the profiler could have more complicated interactions.
> 
> Can you provide a timeline of the `LayerNormLinear` backward with `TORCH_NCCL_ENABLE_TIMING=1`? The provided timeline shows `Linear`, which does not overlap its tensor-parallel communication.

yes, it just record the nccl kernel, i correct the image, i will upload timeline later.

**Comment by cos120** [comment]
@timmoon10 any update？😭

**Comment by timmoon10** [comment]
I reproduce this bug in a pure PyTorch script (see https://github.com/pytorch/pytorch/issues/143890#issuecomment-2576337164), so it doesn't seem specific to TE.

**Comment by cos120** [comment]
> I reproduce this bug in a pure PyTorch script (see https://github.com/pytorch/pytorch/issues/143890#issuecomment-2576337164), so it doesn't seem specific to TE.

Thanks for replay, let's discuss in Pytorch. I will close this issue.
